### PR TITLE
fix: add protection in rest service to always publish with timestamp if user doesn't provide it

### PR DESCRIPTION
**Pull request by Ivansete-status**
## Description
Before this PR, the `relay` REST service accepted "publish" events without `timestamp` info.
Therefore, the _Postgres_ database ended up with the following message set:
image

That caused pagination issues within the _Store_ protocol because the `sender_time` attribute (part of the Cursor object) was always set to 0.

## Changes

- [x] [rest relay: avoid publish messages without timestamp information](https://github.com/waku-org/nwaku/commit/8968ee824031ec817f0013d499c17093d04a1d61)
- [x] [archive.nim: prevent archiving messages without timestamp info](https://github.com/waku-org/nwaku/commit/21ea2b9fa1791ee59309b703abc71756b2f4d2c2)

## Issue
This was noticed while dogfooding `status.test` in the context of https://github.com/status-im/infra-status/issues/37

### CODE REVIEW

**Comment by jm-clius** [comment]
Interesting. Can you expand a little bit more on why the `0` timestamps caused issues with pagination? Although the `timestamp` attribute is mandatory in the Waku Network (which is a certain restricted way of using Waku), the raw protocol does not require timestamps to be present: https://rfc.vac.dev/spec/14/#message-attributes
The only issue this caused for Store was how to do the sorting in the archive, but for this we introduced a `storedAt` timestamp which would always be non-zero. As far as I can see, `0` timestamps should not affect cursor creation or lookups? As long as the Index combination remains unique, the value of the fields constituting the cursor should not matter?

**Comment by gabrielmer** [comment]
I think we've discussed before about automatically adding in REST the current timestamp if the user does not set it in the request.
If that's the case, shouldn't we add the timestamp for the user instead of responding with an error? or am I missing something and I'm mixing different things? 😅

**Comment by Ivansete-status** [comment]
> Interesting. Can you expand a little bit more on why the `0` timestamps caused issues with pagination? Although the `timestamp` attribute is mandatory in the Waku Network (which is a certain restricted way of using Waku), the raw protocol does not require timestamps to be present: https://rfc.vac.dev/spec/14/#message-attributes The only issue this caused for Store was how to do the sorting in the archive, but for this we introduced a `storedAt` timestamp which would always be non-zero. As far as I can see, `0` timestamps should not affect cursor creation or lookups? As long as the Index combination remains unique, the value of the fields constituting the cursor should not matter?

Thanks for the comment!
Let me elaborate more.
Let's imagine a _Store_ request that needs a couple of pages and let's consider that all the messages are archived with `sender_time == 0` as shown in the pic in the PR description.

In this case, the very first page will be returned with the following cursor definition (notice that `senderTime` is zero in the cursor .)
https://github.com/waku-org/nwaku/blob/89dbbd2048d91a97e8f4fefd439f45aef3e9a024/waku/node/waku_node.nim#L807

After that, a go-_Store_ client will ask for the second page of messages from the following code. Notice that the `senderTime` will be zero in this case too:

https://github.com/waku-org/go-waku/blob/16d59f37d02a8375466c027f876acd05621bc9f6/waku/v2/protocol/store/waku_store_client.go#L457-L459

https://github.com/waku-org/go-waku/blob/16d59f37d02a8375466c027f876acd05621bc9f6/waku/v2/protocol/store/pb/store.pb.go#L196

https://github.com/waku-org/go-waku/blob/16d59f37d02a8375466c027f876acd05621bc9f6/waku/v2/protocol/store/pb/store.pb.go#L126

After that, the following `if` statement will happen, throwing an exception, in `nwaku`, and therefore the rest of the _Store_ requests won't proceed. In other words, only the very first page is delivered to the store client.

https://github.com/waku-org/nwaku/blob/c86dc4422f87e6ef0ca0585114762aa60ea76bd6/waku/waku_store/rpc_codec.nim#L56-L57

File: tests/waku_archive/test_waku_archive.nim [hunk 1]
```diff
@@ -72,7 +72,7 @@ suite "Waku Archive - message handling":
 
     ## Then
     check:
-      (waitFor driver.getMessagesCount()).tryGet() == 1
+      (waitFor driver.getMessagesCount()).tryGet() == 0
```
[the remark this answers is not in the document]
Comment (Ivansete-status) [reply tests/waku_archive/test_waku_archive.nim:75]: > I would change the test name. also it should be `deliver` not `driver` 🤣

Good point! I changed the whole tests' names, which is a little bit beyond the scope of this PR.

File: waku/waku_archive/archive.nim [hunk 2]
```diff
@@ -46,7 +46,7 @@ type DefaultMessageValidator* = ref object of MessageValidator
 
 method validate*(validator: DefaultMessageValidator, msg: WakuMessage): ValidationResult =
   if msg.timestamp == 0:
-    return ok()
+    return err(timestampIsZero)
```
Comment (jm-clius) [inline waku/waku_archive/archive.nim:49]: Although I agree with the REST API adding timestamps on behalf of the application when _publishing_ messages. I still don't think the Store should (for now) enforce sender timestamps. Doing so goes against the existing RFC. Furthermore, this obsoletes the entire `receiverTimestamp`, `storedAt`, etc. mechanism which is a larger cleanup exercise. We may indeed decide to follow this route (and perhaps should have from the beginning), but deprecating the RFC provision for optional timestamps is a bit larger, involves an RFC change and a more extensive review of the existing data scheme.

File: waku/waku_api/rest/filter/types.nim [hunk 3]
```diff
@@ -66,7 +66,11 @@ proc toWakuMessage*(msg: FilterWakuMessage, version = 0): Result[WakuMessage, st
     payload = ?msg.payload.decode()
     contentTopic = msg.contentTopic.get(DefaultContentTopic)
     version = uint32(msg.version.get(version))
-    timestamp = msg.timestamp.get(0)
+
+  var timestamp = msg.timestamp.get(0)
+
+  if timestamp == 0:
+    timestamp = getNanosecondTime(getTime().toUnixFloat())
```
[the remark this answers is not in the document]
Comment (Ivansete-status) [reply waku/waku_api/rest/filter/types.nim:73]: Well, this was thought to be a change aligned to the _Store_ change. I can remove it if that doesn't make sense to you in this PR.

**Review by jm-clius — APPROVED** [review]
Thanks! LGTM now.

**Comment by Ivansete-status** [comment]
After a chat with @jm-clius , we left the waku_archive/archive.nim with the following:
```
method validate*(validator: DefaultMessageValidator, msg: WakuMessage): ValidationResult =
  if msg.timestamp == 0:
    return ok()
```

Therefore, messages with timestamp zero are allowed. Kindly ask Hanno for further details as to why this is interesting as he has a better understanding of the system :D

**Pull request merged**